### PR TITLE
refactor(api): use typed keysets

### PR DESCRIPTION
**Pull request by bfredl**
Initially this is just for geting rid of boilerplate, but eventually the types could get exposed as metadata

### CODE REVIEW

**Comment by lewis6991** [comment]
This should automagically work with  https://github.com/neovim/neovim/pull/24523

**Comment by lewis6991** [comment]
I was talking specifically about the changes in `keysets.h` which now adds more specific types to options.

**Comment by bfredl** [comment]
VALIDATE_T is being _moved_ not removed.

File: src/nvim/api/private/helpers.h [hunk 1]
```diff
@@ -66,6 +66,10 @@
 // currently treat key=vim.NIL as if the key was missing
 #define HAS_KEY(o) ((o).type != kObjectTypeNil)
 
+#define HAS_KEY2(d, typ, key) ((d->is_set_ & (1 << KEYSET_OPTIDX_ ## typ ## __ ## key)) != 0)
+
+#define GET_BOOL_DEFAULT_TRUE(d, typ, key) (HAS_KEY2(d, typ, key) ? d->key : true)
```
Comment (justinmk) [inline src/nvim/api/private/helpers.h:71]:
```suggestion
#define GET_BOOL_OR_TRUE(d, typ, key) (HAS_KEY2(d, typ, key) ? d->key : true)
```

File: scripts/gen_eval_files.lua [hunk 2]
```diff
@@ -264,15 +264,15 @@ local function get_api_keysets()
   local keysets = metadata.keysets
 
   for _, keyset in ipairs(keysets) do
-    local kname = keyset[1]
-    local kdef = keyset[2]
-    for _, field in ipairs(kdef) do
-      field[2] = api_type(field[2])
+    local name, keys, types = unpack(keyset)
```
Comment (lewis6991) [inline scripts/gen_eval_files.lua:267]: LuaLs can't analyse this properly so I usually write this as:

```suggestion
    local name, keys, types = keyset[1], keyset[2], keyset[3]
```

Comment (bfredl) [reply scripts/gen_eval_files.lua:267]: shouldn't luals be fixed then? This is a fundamental pattern of the language.

Comment (bfredl) [reply scripts/gen_eval_files.lua:267]: Yes, so we don't do that.  Not using unpack negates the point of using tuple packing in the first place. Where more structure is needed I would switch to named keys instead.

Comment (lewis6991) [reply scripts/gen_eval_files.lua:267]: This is clearly a hill you're going to die on.

Go ahead and disregard my comment.

File: src/nvim/api/options.c [hunk 3]
```diff
@@ -26,14 +27,11 @@
 static int validate_option_value_args(Dict(option) *opts, char *name, int *scope, int *opt_type,
                                       void **from, char **filetype, Error *err)
 {
-  if (HAS_KEY(opts->scope)) {
-    VALIDATE_T("scope", kObjectTypeString, opts->scope.type, {
-      return FAIL;
-    });
-
-    if (!strcmp(opts->scope.data.string.data, "local")) {
+#define HAS_KEY_X(d, v) HAS_KEY(d, option, v)
```
Comment (zeertzjq) [inline src/nvim/api/options.c:30]: Does there need to be an `#undef HAS_KEY_X` somewhere below?

File: src/nvim/api/options.c [hunk 4]
```diff
@@ -44,51 +42,40 @@ static int validate_option_value_args(Dict(option) *opts, char *name, int *scope
...
+  VALIDATE((!HAS_KEY_X(opts, scope) || !HAS_KEY_X(opts, buf)), "%s",
            "cannot use both 'scope' and 'buf'", {
     return FAIL;
   });
 
-  VALIDATE((!HAS_KEY(opts->win) || !HAS_KEY(opts->buf)), "%s", "cannot use both 'buf' and 'win'", {
+  VALIDATE((!HAS_KEY_X(opts, win) || !HAS_KEY_X(opts,
+                                                buf)), "%s", "cannot use both 'buf' and 'win'", {
     return FAIL;
   });
```
Comment (zeertzjq) [inline src/nvim/api/options.c:81]:
```suggestion
#undef HAS_KEY_X

```

File: src/nvim/api/options.c [hunk 5]
```diff
@@ -44,51 +42,40 @@ static int validate_option_value_args(Dict(option) *opts, char *name, int *scope
...
```
Comment (zeertzjq) [inline src/nvim/api/options.c:80]:
```suggestion
  });
  #undef HAS_KEY_X
```

File: src/nvim/api/private/helpers.h [hunk 6]
```diff
@@ -63,8 +63,9 @@
 #define NIL ((Object)OBJECT_INIT)
 #define NULL_STRING ((String)STRING_INIT)
 
-// currently treat key=vim.NIL as if the key was missing
-#define HAS_KEY(o) ((o).type != kObjectTypeNil)
+#define HAS_KEY(d, typ, key) (((d)->is_set_ & (1 << KEYSET_OPTIDX_##typ##__##key)) != 0)
```
Comment (zeertzjq) [inline src/nvim/api/private/helpers.h:66]:
```suggestion
#define HAS_KEY(d, typ, key) (((d)->is_set_ & (1LL << KEYSET_OPTIDX_##typ##__##key)) != 0)
```

File: src/nvim/lua/converter.c [hunk 7]
```diff
@@ -871,7 +871,8 @@ static inline LuaTableProps nlua_check_type(lua_State *const lstate, Error *cons
 {
   if (lua_type(lstate, -1) != LUA_TTABLE) {
     if (err) {
-      api_set_error(err, kErrorTypeValidation, "Expected lua table");
+      api_set_error(err, kErrorTypeValidation, "Expected lua %s",
```
Comment (justinmk) [inline src/nvim/lua/converter.c:874]:
```suggestion
      api_set_error(err, kErrorTypeValidation, "Expected Lua %s",
```

File: src/nvim/lua/converter.c [hunk 8]
```diff
@@ -884,7 +885,7 @@ static inline LuaTableProps nlua_check_type(lua_State *const lstate, Error *cons
 
   if (table_props.type != type) {
     if (err) {
-      api_set_error(err, kErrorTypeValidation, "Unexpected type");
+      api_set_error(err, kErrorTypeValidation, "Expected %s-like lua table", api_typename(type));
```
Comment (justinmk) [inline src/nvim/lua/converter.c:888]:
```suggestion
      api_set_error(err, kErrorTypeValidation, "Expected %s-like Lua table", api_typename(type));
```

File: src/nvim/api/private/helpers.c [hunk 9]
```diff
@@ -915,17 +916,84 @@ HlMessage parse_hl_msg(Array chunks, Error *err)
...
+    } else if (field->type == kObjectTypeBuffer || field->type == kObjectTypeWindow
+               || field->type == kObjectTypeTabpage) {
+      if (value->type == kObjectTypeInteger || value->type == field->type) {
+        *(handle_T *)mem = (handle_T)value->data.integer;
+      } else {
+        api_err_exp(err, field->str, api_typename(field->type), api_typename(value->type));
+        return false;
+      }
+    } else if (field->type == kObjectTypeLuaRef) {
+      api_set_error(err, kErrorTypeValidation, "Invalid key: '%.*s' is only allowed from lua",
```
Comment (zeertzjq) [inline src/nvim/api/private/helpers.c:991]:
```suggestion
      api_set_error(err, kErrorTypeValidation, "Invalid key: '%.*s' is only allowed from Lua",
```

**Comment by clason** [comment]
Let me guess: you're on an old version of LuaLS ;)

(Current version complains about missing table fields if they're not marked optional. Now my config is full of warnings... 😭 )

File: src/nvim/generators/gen_api_dispatch.lua [hunk 10]
```diff
@@ -558,6 +579,7 @@ local function process_function(fn)
   static int %s(lua_State *lstate)
   {
     Error err = ERROR_INIT;
+    char *err_param = 0;
```
[the remark this answers is not in the document]
Comment (bfredl) [reply src/nvim/generators/gen_api_dispatch.lua:582]: yes, coverage analyzing templated code post-monomorphization is expected to contain dead code false positives.

**Pull request merged**